core/src/main/kotlin/com/katanox/tabour/Tabour.kt

```diff
@@ -0,0 +1,88 @@
+package com.katanox.tabour
+
+import com.katanox.tabour.configuration.Registry
+import com.katanox.tabour.consumption.Config
+import com.katanox.tabour.sqs.SqsRegistry
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.newFixedThreadPoolContext
+
+/**
+ * A container of a collection of [Registry]. This class is the entrypoint for interacting with
+ * consumers and producers.
+ *
+ * Consumers and producers are added to a [Registry] and then the registry itself must be added to a
+ * tabour container
+ *
+ * After registering all registries, the tabour container can be started using [start].
+ *
+ * To stop the consumption of the container, you can use [stop]
+ *
+ * By default, the container will allocate 4 threads which can be configured using [Configuration]
+ */
+class Tabour internal constructor(configuration: Configuration) {
+    private val registries: MutableSet<Registry<*>> = mutableSetOf()
+
+    @OptIn(DelicateCoroutinesApi::class)
+    private val dispatcher: CoroutineDispatcher =
+        newFixedThreadPoolContext(configuration.numOfThreads, "tabour")
+
+    private var consumptionStarted: Boolean = false
+
+    private val scope = CoroutineScope(dispatcher)
+
+    /**
+     * Adds a new registry to the Tabour Container
+     *
+     * All registries must be registered before starting the tabour container.
+     */
+    fun <T> register(registry: Registry<T>): Tabour = this.apply { registries.add(registry) }
+
+    /**
+     * Uses a registered SqsProducer to produce a message. The producer must be part of a sqs
+     * registry that has been itself registered
+     *
+     * [registryKey]: the key of the registry which the producer is part of
+     *
+     * [producerKey]: the key of the producer itself
+     *
+     * [produceFn]: A function that returns a Pair<String?, String>. The first part of the pair is
+     * the body of the message and the second part is the message group id. If the body is null, a
+     * message is not produced
+     *
+     * Note: If the registry is not found (either wrong Registry or Producer key), nothing happens.
+     */
+    suspend fun <T, K> produceSqsMessage(
```

Comment:
Question: Based on this, for each integration that we support we will need to add a new function, let's say for Kafka as well?

Reply:
Yes thats correct

Reply:
Doesn't make sense to keep the producer generic as produce messages, and since we are already passing the registry key we can then you can remove the casting the you are doing now, as the logic would still be the same for any other producer, or am I missing something here?
```suggestion
    suspend fun <T, K> produceMessage(
        registryKey: K,
        producerKey: T,
        produceFn: () -> Pair<String?, String>
    ) {
        if (val registry = registries.find { it.key == registryKey }) {
             scope.launch { coroutineScope { registry.produce(producerKey, produceFn) } }
            else -> {}
        }
    }
```

Reply:
the producefn is specific for sqs, because it returns a pair (now changed it to a class) of a body which is string and a message group id.

The produce fn essentially is not valid for kafka for example.

`produceSqsMessage` essentially tells tabour to produce a message using sqs, and you give the sqs producer key to find it (with the registry key)


Making this generic can not work, because then `registry.produce` won't be aware of the concrete type of the `produceFn`

Reply:
But that is solvable with generic, isn't it?